File: frontend/projects/webapp/src/app/pattern/savings-goal-picker/savings-goal-picker-field.ts
import {
  ChangeDetectionStrategy,
  Component,
  computed,
  inject,
  input,
  output,
  signal,
} from '@angular/core';
import { takeUntilDestroyed, toObservable } from '@angular/core/rxjs-interop';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatSelectModule } from '@angular/material/select';
import { MatButtonModule } from '@angular/material/button';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { TranslocoPipe } from '@jsverse/transloco';
import { filter, map } from 'rxjs';
import { cachedResource } from 'ngx-ziflux';
import { formatDate } from 'date-fns';
import {
  getBudgetPeriodForDate,
  parseIsoDateLocal,
  periodIndex,
  WITHDRAWAL_BALANCE_TOLERANCE,
  type BudgetPeriod,
} from 'pulpe-shared';

import { SavingsGoalApi } from '@core/savings-goal/savings-goal-api';
import { AppCurrencyPipe } from '@core/currency';
import { UserSettingsStore } from '@core/user-settings';
import { dateFnsLocaleFor } from '@core/locale';

/**
 * Reusable "Objectif" picker.
 *
 * Two modes, deliberately distinct:
 * - `link` (default) — the 3 CA26 saving-line surfaces: which goal a saving
 *   forecast contributes to. Money goes IN.
 * - `withdrawal` (PUL-329) — which goal funds an income. Money goes OUT, so the
 *   list is restricted to goals with a positive balance and the field shows what
 *   is left after the withdrawal.
 *
 * Value-based (not a Signal-Forms field): the caller passes the current
 * `savingsGoalId` via `[value]` and reacts to `(valueChanged)`. In `link` mode a
 * first option maps to `null` ("Aucun objectif"); in `withdrawal` mode the
 * choice is required, so there is no such option.
 *
 * PUL-313 — when the caller supplies the budget's period, goals whose deadline
 * falls before it are listed but disabled: the `enforce_savings_goal_line_link`
 * trigger would reject the link. Listed, not hidden — a goal that silently
 * disappears is unexplainable. Template lines carry no period and stay
 * unfiltered; the trigger only bounds `budget_line`. Withdrawal callers supply
 * no period either: taking money out of a goal is not bound by its deadline.
 */
@Component({
  selector: 'pulpe-savings-goal-picker-field',
  changeDetection: ChangeDetectionStrategy.OnPush,
  host: { class: 'block' },
  imports: [
    MatFormFieldModule,
    MatSelectModule,
    MatButtonModule,
    MatProgressSpinnerModule,
    TranslocoPipe,
    AppCurrencyPipe,
  ],
  template: `
    @if (isLoading()) {
      <div
        class="flex items-center gap-2 py-2 text-body-small text-on-surface-variant"
        role="status"
        data-testid="savings-goal-picker-loading"
      >
        <mat-progress-spinner mode="indeterminate" [diameter]="20" />
        {{ 'common.loading' | transloco }}
      </div>
    } @else if (error()) {
      <div
        class="flex items-center justify-between gap-3 py-2 text-body-small text-error"
        role="alert"
        data-testid="savings-goal-picker-error"
      >
        <span>{{ 'savingsGoals.loadError' | transloco }}</span>
        <button
          matButton="text"
          type="button"
          (click)="reloadGoals()"
          data-testid="savings-goal-picker-retry"
        >
          {{ 'common.retry' | transloco }}
        </button>
      </div>
    } @else if (mode() === 'withdrawal') {
      <mat-form-field
        appearance="outline"
        subscriptSizing="dynamic"
        class="w-full"
      >
        <mat-label>{{
          'savingsGoals.withdrawalSourceLabel' | transloco
        }}</mat-label>
        <mat-select
          required
          [value]="value()"
          (selectionChange)="valueChanged.emit($event.value)"
          data-testid="savings-goal-withdrawal-select"
        >
          @for (option of withdrawalOptions(); track option.goalId) {
            <mat-option [value]="option.goalId">
              <span>{{ option.name }}</span>
              <span class="ph-no-capture text-on-surface-variant">
                ·
                {{
                  option.availableAmount
                    | appCurrency: option.currency : '1.0-0'
                }}
              </span>
            </mat-option>
          }
        </mat-select>
      </mat-form-field>
      @if (withdrawalOptions().length === 0) {
        <p
          class="mt-1 text-body-small text-on-surface-variant"
          data-testid="savings-goal-withdrawal-empty"
        >
          {{ 'savingsGoals.withdrawalNoFundedGoal' | transloco }}
        </p>
      } @else if (selectedOption(); as option) {
        <p
          class="mt-1 text-body-small text-on-surface-variant ph-no-capture"
          data-testid="savings-goal-withdrawal-preview"
        >
          {{ option.name }} ·
          {{ option.availableAmount | appCurrency: option.currency : '1.0-0' }}
          →
          {{ remainingAmount() | appCurrency: option.currency : '1.0-0' }}
        </p>
        @if (hasInsufficientBalance()) {
          <p
            class="mt-1 text-body-small text-error"
            role="alert"
            data-testid="savings-goal-withdrawal-insufficient"
          >
            {{ 'savingsGoals.withdrawalInsufficient' | transloco }}
          </p>
        } @else if (option.status === 'COMPLETED') {
          <p
            class="mt-1 text-body-small text-on-surface-variant"
            data-testid="savings-goal-withdrawal-completed-note"
          >
            {{ 'savingsGoals.withdrawalCompletedNote' | transloco }}
          </p>
        }
      } @else {
        <!--
          Des objectifs sont proposés, aucun n'est retenu : le bouton d'envoi
          est grisé et c'était le seul état bloqué qui n'en disait pas la
          raison. Couvre aussi l'identifiant absent de la liste rechargée.
        -->
        <p
          class="mt-1 text-body-small text-on-surface-variant"
          data-testid="savings-goal-withdrawal-required"
        >
          {{ 'savingsGoals.withdrawalPickRequired' | transloco }}
        </p>
      }
    } @else {
      <mat-form-field
        appearance="outline"
        subscriptSizing="dynamic"
        class="w-full"
      >
        <mat-label>{{ 'savingsGoals.pickerLabel' | transloco }}</mat-label>
        <mat-select
          [value]="value()"
          (selectionChange)="onSelectionChange($event.value)"
          data-testid="savings-goal-picker-select"
        >
          <mat-option [value]="null">{{
            'savingsGoals.pickerNone' | transloco
          }}</mat-option>
          @for (option of goalOptions(); track option.id) {
            <mat-option
              [value]="option.id"
              [disabled]="option.deadlineLabel !== null"
              [attr.data-testid]="'savings-goal-picker-option-' + option.id"
            >
              {{ option.name }}
              @if (option.deadlineLabel; as deadline) {
                <span class="block text-body-small text-on-surface-variant">
                  {{
                    'savingsGoals.pickerOutsideHorizon'
                      | transloco: { month: deadline }
                  }}
                </span>
              }
            </mat-option>
          }
        </mat-select>
      </mat-form-field>
      @if (goals().length === 0) {
        <p
          class="mt-1 text-body-small text-on-surface-variant"
          data-testid="savings-goal-picker-empty"
        >
          {{ 'savingsGoals.pickerEmpty' | transloco }}
        </p>
      }
    }
  `,
})
export class SavingsGoalPickerField {
  readonly value = input<string | null>(null);
  /** Budget period the line will live in. Omitted on template lines. */
  readonly budgetPeriod = input<BudgetPeriod | null>(null);
  readonly valueChanged = output<string | null>();
  readonly mode = input<'link' | 'withdrawal'>('link');
  /**
   * Withdrawal mode only: the amount actually taken out, already converted into
   * the account currency (RG-009). The original typed amount would compare a
   * foreign-currency figure against a balance held in the account currency.
   */
  readonly withdrawalAmount = input<number | null>(null);

  readonly #api = inject(SavingsGoalApi);
  readonly #settings = inject(UserSettingsStore);

  // Shares the SavingsGoalApi DataCache (key ['savings-goals','list']) with
  // SavingsGoalStore: dedups the fetch across pickers/list and picks up store
  // invalidations.
  readonly #goalsResource = cachedResource({
    cache: this.#api.cache,
    cacheKey: ['savings-goals', 'list'],
    params: () => (this.mode() === 'link' ? {} : undefined),
    loader: () => this.#api.getAll$().pipe(map((r) => r.data ?? [])),
  });

  // Server-filtered: only goals whose confirmed balance is positive, whatever
  // their status. The client never rebuilds that eligibility.
  readonly #withdrawalOptionsResource = cachedResource({
    cache: this.#api.cache,
    cacheKey: ['savings-goals', 'withdrawal-options'],
    params: () => (this.mode() === 'withdrawal' ? {} : undefined),
    loader: () => this.#api.getWithdrawalOptions$().pipe(map((r) => r.data)),
  });

  protected readonly goals = computed(() => this.#goalsResource.value() ?? []);
  protected readonly withdrawalOptions = computed(
    () => this.#withdrawalOptionsResource.value() ?? [],
  );

  protected readonly selectedOption = computed(
    () =>
      this.withdrawalOptions().find(
        (option) => option.goalId === this.value(),
      ) ?? null,
  );

  protected readonly remainingAmount = computed(() => {
    const option = this.selectedOption();
    if (!option) return 0;
    return option.availableAmount - (this.withdrawalAmount() ?? 0);
  });

  // Même bande que le serveur : il accepte `debit <= disponible + tolérance`.
  // Plus serré ici, le pré-contrôle refuserait des retraits que la requête
  // aurait acceptés — vider un pot exactement, d'abord.
  readonly hasInsufficientBalance = computed(
    () => this.remainingAmount() < -WITHDRAWAL_BALANCE_TOLERANCE,
  );

  /**
   * Withdrawal mode: the selection cannot be committed yet — options are still
   * loading or failed, the converted amount is not resolved, or the amount is
   * over the balance. The backend stays the authority; this only spares the user
   * a round-trip.
   */
  readonly isWithdrawalBlocked = computed(() => {
    if (this.mode() !== 'withdrawal') return false;
    if (this.isLoading() || this.error()) return true;
    if (this.value() === null) return true;
    if (this.selectedOption() === null) return true;
    return this.withdrawalAmount() === null || this.hasInsufficientBalance();
  });

  protected readonly isLoading = computed(() =>
    this.mode() === 'withdrawal'
      ? this.#withdrawalOptionsResource.isInitialLoading()
      : this.#goalsResource.isInitialLoading(),
  );
  protected readonly error = computed(() =>
    this.mode() === 'withdrawal'
      ? this.#withdrawalOptionsResource.error()
      : this.#goalsResource.error(),
  );

  /**
   * `deadlineLabel` is non-null exactly when the goal is out of horizon: it
   * both disables the option and names the month that puts it out of reach.
   * Mirrors the trigger's own arithmetic via the shared period calculator —
   * an undated goal has no horizon, so it is never out of it.
   */
  protected readonly goalOptions = computed(() => {
    const period = this.budgetPeriod();
    const payDay = this.#settings.payDayOfMonth();
    const locale = dateFnsLocaleFor(this.#settings.currency());

    return this.goals().map((goal) => {
      if (!period || !goal.targetDate) {
        return { id: goal.id, name: goal.name, deadlineLabel: null };
      }
      const deadline = getBudgetPeriodForDate(
        parseIsoDateLocal(goal.targetDate),
        payDay,
      );
      const isOutsideHorizon = periodIndex(period) > periodIndex(deadline);
      return {
        id: goal.id,
        name: goal.name,
        deadlineLabel: isOutsideHorizon
          ? formatDate(
              new Date(deadline.year, deadline.month - 1, 1),
              'MMMM yyyy',
              { locale },
            )
          : null,
      };
    });
  });

  /**
   * Set once the user picks in THIS picker, as opposed to the link the caller
   * handed us. The distinction decides who may be auto-unlinked below.
   */
  readonly #pickedHere = signal(false);

  protected onSelectionChange(goalId: string | null): void {
    this.#pickedHere.set(true);
    this.valueChanged.emit(goalId);
  }

  /**
   * True when the selection must be withdrawn. Disabling the option is not
   * enough on its own: the caller can widen the period AFTER a goal was picked
   * (a spread range extended past the deadline), leaving a stale id that would
   * still submit. Reuses `goalOptions` — the very list the template disables
   * from — so nothing shown as unselectable can survive as a selection.
   *
   * The two reasons are NOT symmetric. A goal that vanished can never be saved
   * again, so it goes whoever chose it. A goal that is merely out of horizon
   * was legitimately linkable when the line was saved, and an edit surface
   * opens carrying it — dropping that on open would edit the user's data for
   * them, so only a pick made here is taken back.
   *
   * The raw resource value gates the whole thing: the option list collapses to
   * `[]` while loading or errored, and clearing on that would wipe a valid pick.
   */
  readonly #hasStaleSelection = computed(() => {
    if (
      this.#goalsResource.error() ||
      this.#goalsResource.value() === undefined
    )
      return false;
    const selectedId = this.value();
    if (selectedId === null) return false;
    const selected = this.goalOptions().find(
      (option) => option.id === selectedId,
    );
    if (!selected) return true;
    return selected.deadlineLabel !== null && this.#pickedHere();
  });

  constructor() {
    toObservable(this.#hasStaleSelection)
      .pipe(filter(Boolean), takeUntilDestroyed())
      .subscribe(() => this.valueChanged.emit(null));
  }

  protected reloadGoals(): void {
    if (this.mode() === 'withdrawal') {
      this.#withdrawalOptionsResource.reload();
      return;
    }
    this.#goalsResource.reload();
  }
}

File: ios/Pulpe/Shared/Components/SavingsGoalPickerField.swift
import SwiftUI

/// Form field that tags a saving prévision to a savings goal (PUL-12).
///
/// Reused by the template-line editor (primary tagging surface) and the
/// budget-line Add/Edit sheets. Callers show it only for `kind == .saving`;
/// `selection` is the goal id (`nil` = "Aucun objectif"). Reads goals from the
/// app-level `SavingsGoalStore` and refreshes them when it appears.
///
/// Two deliberately distinct modes:
/// - `.link` (default) — which goal a saving forecast feeds. Money goes IN.
/// - `.withdrawal` (PUL-329) — which goal funds an income. Money goes OUT, so
///   the list is the server-filtered set of funded goals and the field shows
///   what is left afterwards. There is no "Aucun objectif": once the caller
///   opted in, the choice is required.
/// - `.plannedWithdrawal` (PUL-329 v2) — which goal an income FORECAST announces
///   a withdrawal from. Nothing leaves the pot yet, so every goal is listed
///   whatever it holds today, and the preview reads the projection of the
///   budget's own month rather than the current balance.
///
/// PUL-313 — when the caller passes `budgetPeriod`, goals whose deadline falls
/// before it are listed but disabled: `enforce_savings_goal_line_link` would
/// reject the link with a 422. Listed, not hidden — a goal that silently
/// disappears is unexplainable. Template lines pass no period and stay
/// unfiltered; the trigger's horizon branch only bounds `budget_line`.
/// `budgetPeriod` can move under a live selection (a spread window extended past
/// a goal's deadline), so the selection is reconciled against it, not just the
/// goal list. A withdrawal passes no period: taking money out of a goal is not
/// bound by the deadline that bounds paying into it.
struct SavingsGoalPickerField: View {
    enum Mode {
        case link
        case withdrawal
        case plannedWithdrawal

        /// Both withdrawal modes read the goals list rather than the balance-bearing
        /// options, and neither is bound by the deadline that bounds paying in.
        var usesGoalsList: Bool { self != .withdrawal }
    }

    /// Local pre-check of a withdrawal choice, kept as a value so it can be
    /// reasoned about (and tested) without a view. The backend stays the
    /// authority; this only spares the user a round-trip.
    struct WithdrawalState: Equatable {
        let selectedOption: SavingsGoalWithdrawalOption?
        /// Already converted into the account currency (RG-009): comparing the
        /// typed amount would weigh a foreign-currency figure against a balance
        /// held in another one.
        let withdrawalAmount: Decimal?
        let isLoading: Bool
        let hasError: Bool

        /// What the goal holds once this income is taken out.
        var remainingAmount: Decimal? {
            guard let selectedOption, let withdrawalAmount else { return nil }
            return selectedOption.availableAmount - withdrawalAmount
        }

        /// Même bande que le serveur, qui accepte `debit <= disponible +
        /// tolérance`. Plus serré ici, le pré-contrôle refuserait des retraits
        /// que la requête aurait acceptés — vider un pot exactement, d'abord.
        var hasInsufficientBalance: Bool {
            guard let remainingAmount else { return false }
            return remainingAmount < -SavingsGoalProgress.withdrawalBalanceTolerance
        }

        var isReady: Bool {
            guard !isLoading, !hasError else { return false }
            guard selectedOption != nil, withdrawalAmount != nil else { return false }
            return !hasInsufficientBalance
        }
    }

    struct SelectionState: Equatable {
        let hasLoadedOnce: Bool
        let isLoading: Bool
        let hasError: Bool
        let knownGoalIDs: Set<String>
        /// Goals the caller's period can still link to — NOT every known goal. The
        /// period moves under a live selection when a spread window is extended.
        let linkableGoalIDs: Set<String>
        /// Whether the live selection was made in this picker rather than handed in.
        let pickedHere: Bool

        /// The two reasons a selection can go stale are NOT symmetric. A goal that
        /// vanished can never be saved again, so it drops whoever chose it. A goal
        /// that is merely out of horizon was legitimately linkable when the line was
        /// saved, and an edit sheet opens carrying it — withdrawing that on open
        /// would edit the user's data for them, so only a pick made here is taken back.
        func reconciled(_ selection: String?) -> String? {
            guard hasLoadedOnce, !isLoading, !hasError, let selection else {
                return selection
            }
            guard knownGoalIDs.contains(selection) else { return nil }
            if linkableGoalIDs.contains(selection) { return selection }
            return pickedHere ? nil : selection
        }
    }

    @Binding var selection: String?
    var mode: Mode = .link
    var budgetPeriod: BudgetPeriod?
    /// Withdrawal mode only: the amount actually taken out, already converted
    /// into the account currency (RG-009). Comparing the typed amount would
    /// weigh a foreign-currency figure against a balance held in another one.
    var withdrawalAmount: Decimal?
    /// Bump to re-read the options after the backend refused a stale balance.
    var withdrawalRefreshToken = 0
    /// Fires whenever the local checks agree the selection could be submitted.
    var onWithdrawalReadinessChange: (Bool) -> Void = { _ in }

    @State private var pickedHere = false

    @Environment(SavingsGoalStore.self) private var store
    @Environment(UserSettingsStore.self) private var userSettingsStore

    /// The deadline period a goal puts a saving out of reach past, or `nil` when
    /// the link is allowed. Mirrors the trigger's own arithmetic through the
    /// shared period port — an undated goal has no horizon to fall outside of,
    /// and neither does a template line, which carries no period at all.
    static func exceededDeadline(
        for goal: SavingsGoal,
        budgetPeriod: BudgetPeriod?,
        payDayOfMonth: Int?
    ) -> BudgetPeriod? {
        guard let budgetPeriod, let targetDate = goal.targetDateValue else { return nil }
        let deadline = BudgetPeriodCalculator.periodForDate(targetDate, payDayOfMonth: payDayOfMonth)
        return BudgetPeriodCalculator.comparePeriods(budgetPeriod, deadline) > 0 ? deadline : nil
    }

    private func exceededDeadline(for goal: SavingsGoal) -> BudgetPeriod? {
        Self.exceededDeadline(
            for: goal,
            budgetPeriod: budgetPeriod,
            payDayOfMonth: userSettingsStore.payDayOfMonth
        )
    }

    private var selectedGoal: SavingsGoal? {
        guard let selection else { return nil }
        return store.goals.first { $0.id == selection }
    }

    private var selectionState: SelectionState {
        SelectionState(
            hasLoadedOnce: store.hasLoadedOnce,
            isLoading: store.isLoading,
            hasError: store.error != nil,
            knownGoalIDs: Set(store.goals.map(\.id)),
            linkableGoalIDs: Set(store.goals.filter { exceededDeadline(for: $0) == nil }.map(\.id)),
            pickedHere: pickedHere
        )
    }

    var body: some View {
        VStack(alignment: .leading, spacing: DesignTokens.Spacing.sm) {
            Text(mode == .link ? "Objectif" : "Objectif utilisé")
                .font(PulpeTypography.labelMedium)
                .foregroundStyle(Color.onSurfaceVariant)

            switch mode {
            case .withdrawal:
                SavingsGoalWithdrawalPicker(
                    selection: $selection,
                    withdrawalAmount: withdrawalAmount,
                    refreshToken: withdrawalRefreshToken,
                    onReadinessChange: onWithdrawalReadinessChange
                )
            case .plannedWithdrawal:
                SavingsGoalPlannedWithdrawalPicker(
                    selection: $selection,
                    withdrawalAmount: withdrawalAmount,
                    budgetPeriod: budgetPeriod,
                    onReadinessChange: onWithdrawalReadinessChange
                )
            case .link:
                pickerContent
            }
        }
        .task {
            guard mode.usesGoalsList else { return }
            await store.loadIfNeeded()
        }
        .onChange(of: selectionState, initial: true) { _, state in
            guard mode == .link else { return }
            selection = state.reconciled(selection)
        }
    }

    @ViewBuilder
    private var pickerContent: some View {
        if store.hasError {
            errorContent
        } else if !store.hasLoadedOnce {
            loadingContent
        } else if store.goals.isEmpty {
            emptyContent
        } else {
            menuContent
        }
    }

    private var errorContent: some View {
        savingsGoalFieldError { Task { await store.forceRefresh() } }
    }

    private var loadingContent: some View {
        savingsGoalFieldLoading()
    }

    private var emptyContent: some View {
        savingsGoalFieldSurface {
            Text("Aucun objectif disponible")
                .foregroundStyle(Color.onSurfaceVariant)
        }
        .accessibilityLabel("Aucun objectif d'épargne disponible")
    }

    private var menuContent: some View {
        Menu {
            savingsGoalPickerButton(title: AppLocale.string("Aucun objectif"), isSelected: selection == nil) {
                pickedHere = true
                selection = nil
            }
            Divider()
            ForEach(store.goals) { goal in
                if let deadline = exceededDeadline(for: goal) {
                    savingsGoalPickerButton(
                        title: goal.name,
                        subtitle: AppLocale.string(
                            "Échéance dépassée · \(Formatters.monthName(for: deadline.month)) \(deadline.year)"
                        ),
                        isSelected: goal.id == selection
                    ) {}
                    .disabled(true)
                } else {
                    savingsGoalPickerButton(title: goal.name, isSelected: goal.id == selection) {
                        pickedHere = true
                        selection = goal.id
                    }
                }
            }
        } label: {
            savingsGoalFieldSurface {
                Text(selectedGoal?.name ?? AppLocale.string("Aucun objectif"))
                    .foregroundStyle(selectedGoal == nil ? Color.onSurfaceVariant : Color.textPrimary)
                Spacer()
                Image(systemName: "chevron.up.chevron.down")
                    .font(.caption)
                    .foregroundStyle(Color.onSurfaceVariant)
            }
        }
        .accessibilityLabel("Objectif d'épargne")
        .accessibilityValue(selectedGoal?.name ?? AppLocale.string("Aucun objectif"))
    }
}

// MARK: - Withdrawal mode (PUL-329)

/// Presentation of the funded-goal choice. Owns its own fetch: the option list
/// carries balances, which the goals cache does not.
private struct SavingsGoalWithdrawalPicker: View {
    @Binding var selection: String?
    let withdrawalAmount: Decimal?
    let refreshToken: Int
    let onReadinessChange: (Bool) -> Void

    @Environment(SavingsGoalStore.self) private var store
    @State private var options: [SavingsGoalWithdrawalOption] = []
    @State private var isLoading = true
    @State private var error: Error?

    private var state: SavingsGoalPickerField.WithdrawalState {
        SavingsGoalPickerField.WithdrawalState(
            selectedOption: selection.flatMap { id in options.first { $0.goalId == id } },
            withdrawalAmount: withdrawalAmount,
            isLoading: isLoading,
            hasError: error != nil
        )
    }

    var body: some View {
        VStack(alignment: .leading, spacing: DesignTokens.Spacing.xs) {
            content
        }
        .task(id: refreshToken) { await load() }
        .onChange(of: state.isReady, initial: true) { _, isReady in
            onReadinessChange(isReady)
        }
    }

    @ViewBuilder
    private var content: some View {
        if error != nil {
            errorContent
        } else if isLoading {
            loadingContent
        } else if options.isEmpty {
            Text("Aucun objectif n'a d'argent disponible pour l'instant.")
                .font(PulpeTypography.footnote)
                .foregroundStyle(Color.onSurfaceVariant)
        } else {
            menuContent
            preview
        }
    }

    private var errorContent: some View {
        savingsGoalFieldError { Task { await load() } }
    }

    private var loadingContent: some View {
        savingsGoalFieldLoading()
    }

    private var menuContent: some View {
        Menu {
            ForEach(options) { option in
                savingsGoalPickerButton(
                    title: "\(option.name) · \(option.availableAmount.asAdaptiveCurrency(option.currency))",
                    isSelected: option.goalId == selection
                ) {
                    selection = option.goalId
                }
            }
        } label: {
            savingsGoalFieldSurface {
                Text(state.selectedOption?.name ?? AppLocale.string("Choisis un objectif"))
                    .foregroundStyle(state.selectedOption == nil ? Color.onSurfaceVariant : Color.textPrimary)
                Spacer()
                Image(systemName: "chevron.up.chevron.down")
                    .font(.caption)
                    .foregroundStyle(Color.onSurfaceVariant)
            }
        }
        .accessibilityLabel("Objectif utilisé")
        .accessibilityValue(state.selectedOption?.name ?? AppLocale.string("Aucun objectif choisi"))
    }

    @ViewBuilder
    private var preview: some View {
        if let selectedOption = state.selectedOption {
            if let remainingAmount = state.remainingAmount {
                Text(
                    "\(selectedOption.name) · "
                        + "\(selectedOption.availableAmount.asAdaptiveCurrency(selectedOption.currency)) → "
                        + remainingAmount.asAdaptiveCurrency(selectedOption.currency)
                )
                .font(PulpeTypography.footnote)
                .foregroundStyle(Color.onSurfaceVariant)
                .sensitiveAmount()
            }
            if state.hasInsufficientBalance {
                Text("Ce montant dépasse ce que contient l'objectif.")
                    .font(PulpeTypography.footnote)
                    .foregroundStyle(Color.destructivePrimary)
            } else if selectedOption.status == .completed {
                Text("Cet objectif reste atteint. Tu pourras le rouvrir depuis son détail.")
                    .font(PulpeTypography.footnote)
                    .foregroundStyle(Color.onSurfaceVariant)
            }
        }
    }

    private func load() async {
        isLoading = true
        error = nil
        defer { isLoading = false }
        do {
            // A refresh means the displayed balances were refused: bypass the cache.
            options = try await store.fetchWithdrawalOptions(forceRefresh: refreshToken > 0)
            if let selection, !options.contains(where: { $0.goalId == selection }) {
                self.selection = nil
            }
        } catch {
            self.error = error
        }
    }
}

// MARK: - Shared field chrome

/// Internal rather than file-private: `SavingsGoalPlannedWithdrawalPicker` wears
/// the same chrome from its own file, and this file is already at the length
/// budget that made splitting it necessary.
@ViewBuilder
func savingsGoalFieldSurface(
    @ViewBuilder content: () -> some View
) -> some View {
    HStack(spacing: DesignTokens.Spacing.sm) {
        content()
    }
    .padding(.horizontal, DesignTokens.Spacing.lg)
    .frame(maxWidth: .infinity, minHeight: DesignTokens.TapTarget.minimum, alignment: .leading)
    .background(
        Color.surfaceContainerLow,
        in: RoundedRectangle(cornerRadius: DesignTokens.CornerRadius.button)
    )
}

/// Failed load, with the retry the caller owns — the three pickers fetch from
/// three different places but say the same thing when the fetch fails.
/// `@MainActor` because `textLinkButtonStyle()` is: a free function inherits no
/// isolation from the view body that calls it.
@MainActor @ViewBuilder
func savingsGoalFieldError(retry: @escaping () -> Void) -> some View {
    savingsGoalFieldSurface {
        Image(systemName: "exclamationmark.triangle")
            .foregroundStyle(Color.destructivePrimary)
        Text("Impossible de charger les objectifs")
            .font(PulpeTypography.footnote)
            .foregroundStyle(Color.onSurfaceVariant)
        Spacer()
        Button("Réessayer", action: retry)
            .textLinkButtonStyle()
    }
    .accessibilityElement(children: .contain)
}

@ViewBuilder
func savingsGoalFieldLoading() -> some View {
    savingsGoalFieldSurface {
        ProgressView()
            .controlSize(.small)
        Text("Chargement des objectifs…")
            .font(PulpeTypography.footnote)
            .foregroundStyle(Color.onSurfaceVariant)
    }
    .accessibilityElement(children: .combine)
}

/// `subtitle` rides in the same button so the reason travels with the goal
/// it disables — a menu entry cannot carry a separate explanatory row.
@ViewBuilder
func savingsGoalPickerButton(
    title: String,
    subtitle: String? = nil,
    isSelected: Bool,
    action: @escaping () -> Void
) -> some View {
    Button(action: action) {
        if isSelected {
            Label(title, systemImage: "checkmark")
        } else {
            Text(title)
        }
        if let subtitle {
            Text(subtitle)
        }
    }
}
